### PR TITLE
LOOP-1686: add higher level tests to Loop algorithm

### PR DESCRIPTION
**Pull request by novalegra**
https://tidepool.atlassian.net/browse/LOOP-1686

The number of files looks really big because of all the test fixtures. To generate fixtures from issue reports, I created a small script that can be found at https://github.com/tidepool-org/loop-issue-report/tree/novalegra/generate-loop-fixtures

### CODE REVIEW

File: Loop/Managers/WatchDataManager.swift [hunk 1]
```diff
@@ -194,7 +194,7 @@ final class WatchDataManager: NSObject {
             
             if let glucose = glucose {
                 updateGroup.enter()
-                manager.glucoseStore.getCachedGlucoseSamples(start: glucose.startDate) { (samples) in
+                manager.glucoseStore.getCachedGlucoseSamples(start: glucose.startDate, end: nil) { (samples) in
```
Comment (novalegra) [inline Loop/Managers/WatchDataManager.swift]: I had to add all of these `end: nil`s because protocols don't have default arguments

File: LoopTests/LoopDataManagerTests.swift [hunk 2]
```diff
@@ -0,0 +1,300 @@
+//
```
Comment (novalegra) [inline LoopTests/LoopDataManagerTests.swift:1]: Please review this file! (I know it gets hidden among the fixtures)

File: LoopTests/LoopDataManagerTests.swift [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+        )
+        
+        let doseStore = MockDoseStore(for: test)
+        doseStore.basalProfileApplyingOverrideHistory = loadBasalRateScheduleFixture("basal_profile")
+        let glucoseStore = MockGlucoseStore(for: test)
+        let carbStore = MockCarbStore(for: test)
+        
+        let currentDate = glucoseStore.latestGlucose!.startDate
+        
+        loopDataManager = LoopDataManager(
```
Comment (novalegra) [inline LoopTests/LoopDataManagerTests.swift:93]: Initialize the LDM with mock stores that return different effects depending on which `DataManagerTestType` they're set up with

File: LoopTests/Mock Stores/HKHealthStoreMock+LoopTests.swift [hunk 4]
```diff
@@ -0,0 +1,76 @@
+//
```
Comment (novalegra) [inline LoopTests/Mock Stores/HKHealthStoreMock+LoopTests.swift]: Copy-pased directly from LoopKit without any edits

File: LoopTests/Mock Stores/MockCarbStore.swift [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    }
+    
+    func glucoseEffects<Sample>(of samples: [Sample], startingAt start: Date, endingAt end: Date?, effectVelocities: [GlucoseEffectVelocity]?) throws -> [GlucoseEffect] where Sample : CarbEntry {
+        return []
+    }
+    
+    func getCarbsOnBoardValues(start: Date, end: Date?, effectVelocities: [GlucoseEffectVelocity]?, completion: @escaping ([CarbValue]) -> Void) {
+        completion([])
+    }
+
```
Comment (novalegra) [inline LoopTests/Mock Stores/MockCarbStore.swift:82]: I didn't have these functions to return any meaningful values because they aren't used in the portion of the algo that I'm testing

File: LoopTests/Mock Stores/MockCarbStore.swift [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        return []
+    }
+    
+    func getCarbsOnBoardValues(start: Date, end: Date?, effectVelocities: [GlucoseEffectVelocity]?, completion: @escaping ([CarbValue]) -> Void) {
+        completion([])
+    }
+    
+    func carbsOnBoard(at date: Date, effectVelocities: [GlucoseEffectVelocity]?, completion: @escaping (CarbStoreResult<CarbValue>) -> Void) {
+        completion(.failure(.notConfigured))
+    }
```
Comment (novalegra) [inline LoopTests/Mock Stores/MockCarbStore.swift:83]: I didn't have these functions to return any meaningful values because they aren't used in the portion of the algo that I'm testing

File: LoopTests/Mock Stores/MockDoseStore.swift [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    func getInsulinOnBoardValues(start: Date, end: Date?, basalDosingEnd: Date?, completion: @escaping (DoseStoreResult<[InsulinValue]>) -> Void) {
+        completion(.failure(.configurationError))
+    }
+    
+    func getNormalizedDoseEntries(start: Date, end: Date?, completion: @escaping (DoseStoreResult<[DoseEntry]>) -> Void) {
+        completion(.failure(.configurationError))
+    }
+    
+    func executePumpEventQuery(fromQueryAnchor queryAnchor: DoseStore.QueryAnchor?, limit: Int, completion: @escaping (DoseStore.PumpEventQueryResult) -> Void) {
+        completion(.failure(DoseStore.DoseStoreError.configurationError))
```
Comment (novalegra) [inline LoopTests/Mock Stores/MockDoseStore.swift:80]: I didn't have these functions to return any meaningful values because they aren't used in the portion of the algo that I'm testing (but they could be updated later if needed for other LDM tests)

File: LoopTests/Mock Stores/MockGlucoseStore.swift [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+    
+    func executeGlucoseQuery(fromQueryAnchor queryAnchor: GlucoseStore.QueryAnchor?, limit: Int, completion: @escaping (GlucoseStore.GlucoseQueryResult) -> Void) {
+        // Using the dose store error because we don't need to create GlucoseStore errors just for the mock store
+        completion(.failure( DoseStore.DoseStoreError.configurationError))
+    }
+    
+    func counteractionEffects<Sample>(for samples: [Sample], to effects: [GlucoseEffect]) -> [GlucoseEffectVelocity] where Sample : GlucoseSampleValue {
+        return [] // TODO: check if we'll ever want to test this
+    }
+
```
Comment (novalegra) [inline LoopTests/Mock Stores/MockGlucoseStore.swift:80]: I didn't have these functions to return any meaningful values because they aren't used in the portion of the algo that I'm testing (but they could be updated later if needed for other LDM tests)

File: LoopTests/Mock Stores/MockGlucoseStore.swift [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+        case .highAndStable:
+            return 200.0
+        case .highAndRisingWithCOB:
+            return 129.93174411197853
+        case .lowAndFallingWithCOB:
+            return 75.10768374646841
+        case .lowWithLowTreatment:
+            return 81.22399763523448
+        case .highAndFalling:
+            return 200.0
```
Comment (novalegra) [inline LoopTests/Mock Stores/MockGlucoseStore.swift:175]: Because the algorithm only explicitly needs the most recent cached glucose value, I chose to just create that 'last' glucose value (other cached glucose values are used to generate effects, but we're not generating effects in this mock store)

File: Loop/Managers/LoopDataManager.swift [hunk 10]
```diff
@@ -1482,6 +1510,18 @@ extension LoopDataManager {
...
+    public func updateTheLoop(completion: @escaping ([PredictedGlucoseValue]?, (recommendation: TempBasalRecommendation, date: Date)?, (recommendation: BolusRecommendation, date: Date)?) -> Void) {
+        self.dataAccessQueue.async {
+            do {
+                try self.update()
+                completion(self.predictedGlucose, self.recommendedTempBasal, self.recommendedBolus)
+            } catch let error {
+                self.logger.error("%{public}@", String(describing: error))
+            }
+        }
+    }
```
Comment (novalegra) [inline Loop/Managers/LoopDataManager.swift]: I exposed the `update` function for testing purposes. I'm open to changing this, but the reason I chose to do it this way is so I could have a completion handler that would return the parts of the prediction/recommendations that were needed for testing.

File: Loop/Managers/LoopDataManager.swift [hunk 11]
```diff
@@ -1482,6 +1510,19 @@ extension LoopDataManager {
...
+extension LoopDataManager {
+    // This should ONLY be used for testing
+    public func updateAndGetRecommentations(completion: @escaping ([PredictedGlucoseValue]?, (recommendation: TempBasalRecommendation, date: Date)?, (recommendation: BolusRecommendation, date: Date)?) -> Void) {
+        self.dataAccessQueue.async {
+            do {
+                try self.update()
+                completion(self.predictedGlucose, self.recommendedTempBasal, self.recommendedBolus)
+            } catch let error {
+                self.logger.error("%{public}@", String(describing: error))
+            }
```
Comment (novalegra) [inline Loop/Managers/LoopDataManager.swift]: This adds a public way to access the `update` function, which is something to be aware of. I'm open to changing this, but the reason I did it this way is because it allows for a completion handler that can return all the needed information to carry out the LDM effects tests

File: Loop/Managers/LoopDataManager.swift [hunk 12]
```diff
@@ -78,40 +84,54 @@ final class LoopDataManager {
 
         cacheStore = PersistenceController.controllerInAppGroupDirectory()
 
-        carbStore = CarbStore(
```
Comment (darinkrauss) [inline Loop/Managers/LoopDataManager.swift:81]: [nit] I would recommend: 1) change `carbStore` property back to `let` (above), 2) rename `carbStorage` parameter (above) to `carbStore`, and 3) change below code to something like:
```
self.carbStore = carbStore ?? CarbStore(...
```
And the same thing for the two other stores.

File: Loop/Managers/LoopDataManager.swift [hunk 13]
```diff
@@ -322,7 +342,11 @@ final class LoopDataManager {
         NotificationManager.clearLoopNotRunningNotifications()
         NotificationManager.scheduleLoopNotRunningNotifications()
         analyticsServicesManager.loopDidSucceed(duration)
-        storeDosingDecision(withDate: date)
+        // The current date is only set while testing, and we don't want to store the dosing decision while testing
```
Comment (darinkrauss) [inline Loop/Managers/LoopDataManager.swift]: This is trying to solve a very wide ranging problem. (That of persisting data during tests.) I would not special case it here.

File: Loop/Managers/LoopDataManager.swift [hunk 14]
```diff
@@ -40,6 +40,8 @@ final class LoopDataManager {
     private let logger = DiagnosticLog(category: "LoopDataManager")
 
     private let analyticsServicesManager: AnalyticsServicesManager
+    
+    private let test_currentDate: Date?
```
Comment (darinkrauss) [inline Loop/Managers/LoopDataManager.swift]: [nit] I'm honestly not a fan of this testing "hack". However, if we need this can we at least hide it away from most of the code. Perhaps having a private function that returns "now" and uses `test_currentDate` if set, or otherwise `Date()`? That way `test_currentDate` isn't sprinkled throughout the code (and less likely to miss a case).

**Review by darinkrauss — COMMENTED** [review]
Not yet complete, but here are some comments. I'll do another full review after you move the protocols to Loop.

File: Loop/Managers/LoopDataManager.swift (hunk 13 above)
Comment (ps2) [reply Loop/Managers/LoopDataManager.swift]: I also think it would be a good test to assert that dosing decisions *are* stored after a Loop.

File: Loop/Managers/LoopDataManager.swift (hunk 14 above)
Comment (ps2) [reply Loop/Managers/LoopDataManager.swift]: Yes, I agree with @darinkrauss

Comment (rickpasetto) [reply Loop/Managers/LoopDataManager.swift]: The way I've done this in the past is similar to what @darinkrauss suggests above: inject a function that returns a date, and have it's default to return `Date()`. Thusly:

```
private let now: () -> Date
```

 ```
init(
    ...
        now: () -> Date = { Date() }
) {
...
    self.now = now
```

... then in all callers instead of `Date()` just call `now()` (or `self.now()`).
Then unit tests can just inject their own function.  Almost the same thing as above, just thought I'd mention it. :)

File: Loop/Managers/LoopDataManager.swift (hunk 11 above)
Comment (ps2) [reply Loop/Managers/LoopDataManager.swift]: This is testing internal implementation of LDM. Instead, we should be checking for the externally visible effects. For example, in a test we can provide a LoopDataManagerDelegate that listens for the didRecommendBasalChange, to assert any basal changes when a loop is run.

For the other effects there are similar ways of asserting the data. It can be tricky, since much of this asynchronous, and may require some changes to LDM itself, but those changes should be fairly minimal.

File: Loop/Managers/RemoteDataServicesManager.swift [hunk 15]
```diff
@@ -31,21 +31,21 @@ final class RemoteDataServicesManager {
 
     private let log = OSLog(category: "RemoteDataServicesManager")
     
-    private var carbStore: CarbStore
+    private var carbStore: CarbStoreProtocol
```
Comment (ps2) [inline Loop/Managers/RemoteDataServicesManager.swift]: Why are we changing RemoteDataServicesManager to use the protocols?

File: Loop/View Controllers/StatusTableViewController.swift [hunk 16]
```diff
@@ -1112,7 +1112,10 @@ final class StatusTableViewController: LoopChartsTableViewController {
                 vc.restoreUserActivityState(activity)
             }
         case let vc as InsulinDeliveryTableViewController:
-            vc.doseStore = deviceManager.loopManager.doseStore
+            guard let doseStore = deviceManager.loopManager.doseStore as? DoseStore else {
+                fatalError("Mock LoopDataManager should not use UI components")
```
Comment (ps2) [inline Loop/View Controllers/StatusTableViewController.swift]: nit: "Mock LoopDataManager should not use UI components" -> "InsulinDeliveryTableViewController cannot use mocked DoseStore"

File: LoopTests/LoopDataManagerTests.swift [hunk 17]
```diff
@@ -0,0 +1,300 @@
...
+        return TimeZone(secondsFromGMT: 25200)!
+    }
+    
+    static var utcTimeZone: TimeZone {
+        return TimeZone(secondsFromGMT: 0)!
+    }
+}
+
+extension ISO8601DateFormatter {
+    static func localTimeDate(timeZone: TimeZone = .fixtureTimeZone) -> Self {
```
Comment (ps2) [inline LoopTests/LoopDataManagerTests.swift:37]: nit: `localTimeDate` -> `local`, or `localTimeFormatter`

File: LoopTests/Mock Stores/HKHealthStoreMock+LoopTests.swift (hunk 4 above)
Comment (ps2) [reply LoopTests/Mock Stores/HKHealthStoreMock+LoopTests.swift]: I am guessing this is being included because all the reaching into LDM for the HKHealthStore. I think it'd be nice to move the authorization stuff into DeviceDataManager, and keep it out of LDM. Then we wouldn't need to bring the mock over; HK access really should be the domain of the stores, and LDM should't worry about it at all.

**Review by ps2 — CHANGES_REQUESTED** [review]
In general, this looks really good! I think the way the tests are set up, while still a bit unwieldy, is a solid framework for testing Loop's algorithm at this level. I do have some comments on things I think we can improve on; they're not necessarily simple, so let me know if you would like to discuss.

File: Loop/Managers/LoopDataManager.swift (hunk 14 above)
Comment (ps2) [reply Loop/Managers/LoopDataManager.swift]: Yes, we do this in DashPumpManager, as well.

Comment (novalegra) [reply Loop/Managers/LoopDataManager.swift]: Done

File: Loop/Managers/LoopDataManager.swift (hunk 12 above)
Comment (novalegra) [reply Loop/Managers/LoopDataManager.swift:81]: Agreed about (2) and (3), but if the `carbStore` property isn't a var, I get a build error on `carbStore.carbRatioSchedule = newValue` and `carbStore.insulinSensitivitySchedule = newValue`. These lines could be factored out, but I'm not sure what the best course of action is. @ps2 any thoughts?

Comment (novalegra) [reply Loop/Managers/LoopDataManager.swift:81]: Perhaps a delegate to set the carb ratio/insulin sensitivities in `CarbStore`

File: Loop/Managers/RemoteDataServicesManager.swift (hunk 15 above)
Comment (novalegra) [reply Loop/Managers/RemoteDataServicesManager.swift]: `DeviceDataManager` uses `loopManager.carbStore` to init `RemoteDataServicesManager`, but you could force cast here (which I've now implemented)

File: LoopTests/Mock Stores/HKHealthStoreMock+LoopTests.swift (hunk 4 above)
Comment (novalegra) [reply LoopTests/Mock Stores/HKHealthStoreMock+LoopTests.swift]: This is actually being included so the mock stores can be generated (like `MockDoseStore`, etc) for testing LDM. I think we'd need to bring this mock over regardless, unless the store mocks should be located in LoopKit instead of LoopTests

Comment (ps2) [reply LoopTests/Mock Stores/HKHealthStoreMock+LoopTests.swift]: I know the HKHealthStore is referenced in the constructor of the stores, but the constructor isn't part of the protocol. Where else is the HKHealthStore used in the mocks?

Comment (novalegra) [reply LoopTests/Mock Stores/HKHealthStoreMock+LoopTests.swift]: `InsulinDeliveryStore` is part of the `DoseStoreProtocol`, and `InsulinDeliveryStore` requires a `HKHealthStore`. I can mock the `InsulinDeliveryStore` if we want to pull out the `HKHealthStoreMock`; `HKHealthStoreMock` isn't being used in a meaningful way for these tests, so it's not critical that it be there, but I was also hesitant to create another mock if it wasn't necessary

File: Loop/Managers/LoopDataManager.swift (hunk 11 above)
Comment (novalegra) [reply Loop/Managers/LoopDataManager.swift]: I was able to actually just use the `getLoopStatus` to get what I need!

File: LoopTests/Mock Stores/HKHealthStoreMock+LoopTests.swift (hunk 4 above)
Comment (novalegra) [reply LoopTests/Mock Stores/HKHealthStoreMock+LoopTests.swift]: I mocked `InsulinDeliveryStore` and initialized the stores/moved the logic for authorization into the device manager

File: Loop/Managers/DeviceDataManager.swift [hunk 18]
```diff
@@ -67,6 +67,42 @@ final class DeviceDataManager {
             UserDefaults.appGroup?.pumpManagerRawValue = pumpManager?.rawValue
         }
     }
+    
+    // MARK: Stores
+    let healthStore: HKHealthStore
```
Comment (novalegra) [inline Loop/Managers/DeviceDataManager.swift:72]: Moving authorization stuff out of LDM and into device data manager per @ps2's suggestion

File: Loop/Managers/DeviceDataManager.swift [hunk 19]
```diff
@@ -115,6 +151,46 @@ final class DeviceDataManager {
         self.pluginManager = pluginManager
         self.alertManager = alertManager
         
+        self.healthStore = HKHealthStore()
```
Comment (novalegra) [inline Loop/Managers/DeviceDataManager.swift:158]: Initializing the stores in DDM so we can do HK authorization with them (and not to have to mock the HKHealth store for testing)

File: Loop/Managers/RemoteDataServicesManager.swift [hunk 20]
```diff
@@ -33,19 +33,19 @@ final class RemoteDataServicesManager {
     
     private var carbStore: CarbStore
 
-    private var doseStore: DoseStore
+    private var doseStore: DoseStoreProtocol
```
Comment (novalegra) [inline Loop/Managers/RemoteDataServicesManager.swift]: Adding store protocols so the stores can be mocked to test LDM

File: Loop/Managers/Store Protocols/CarbStoreProtocol.swift [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+//  Loop
+//
+//  Created by Anna Quinlan on 8/17/20.
+//  Copyright © 2020 LoopKit Authors. All rights reserved.
+//
+
+import LoopKit
+import HealthKit
+
+public protocol CarbStoreProtocol: AnyObject {
```
Comment (novalegra) [inline Loop/Managers/Store Protocols/CarbStoreProtocol.swift]: For all of these stores, I've checked that all of the properties are needed in order for LDM to not error/function properly. I _could_ do more checking & casting into the desired store type, but I tried to minimize that so I'm not cluttering LDM

File: Loop/View Controllers/CarbAbsorptionViewController.swift [hunk 22]
```diff
@@ -48,7 +48,7 @@ final class CarbAbsorptionViewController: LoopChartsTableViewController, Identif
                     self?.reloadData(animated: true)
                 }
             },
-            notificationCenter.addObserver(forName: .HKUserPreferencesDidChange, object: deviceManager.loopManager.glucoseStore.healthStore, queue: nil) {[weak self] _ in
+            notificationCenter.addObserver(forName: .HKUserPreferencesDidChange, object: deviceManager.glucoseStore.healthStore, queue: nil) {[weak self] _ in
```
Comment (novalegra) [inline Loop/View Controllers/CarbAbsorptionViewController.swift:51]: Going through the device data manager to get to the stores

**Comment by novalegra** [comment]
I've responded to your (amazing) comments, @ps2, @darinkrauss, and @rickpasetto. Major changes were moving the protocols into Loop, refactoring the authorization out of LDM and into DeviceDataManager, creating a mock for `InsulinDeliveryStore`, and passing a closure into LDM that returns the "current" date for the purposes of testing.

File: Loop/Managers/DeviceDataManager.swift [hunk 23]
```diff
@@ -136,15 +212,19 @@ final class DeviceDataManager {
...
+            doseStore: doseStore,
+            glucoseStore: glucoseStore,
+            carbStore: carbStore,
+            cacheStore: cacheStore
         )
         watchManager = WatchDataManager(deviceManager: self)
 
         let remoteDataServicesManager = RemoteDataServicesManager(
-            carbStore: loopManager.carbStore,
-            doseStore: loopManager.doseStore,
```
Comment (ps2) [inline Loop/Managers/DeviceDataManager.swift:145]: All the removals of this "reaching into another object" is a good sign that refactoring stores from LDM to DDM is a good move. Love to see the cleanup!

File: Loop/Managers/DeviceDataManager.swift [hunk 24]
```diff
@@ -721,16 +815,18 @@ extension DeviceDataManager {
         }
 
         let devicePredicate = HKQuery.predicateForObjects(from: [testingPumpManager.testingDevice])
-        let doseStore = loopManager.doseStore
-        let insulinDeliveryStore = doseStore.insulinDeliveryStore
+        guard let insulinDeliveryStore = doseStore.insulinDeliveryStore as? InsulinDeliveryStore else {
```
Comment (ps2) [inline Loop/Managers/DeviceDataManager.swift]: Cast not needed here.

File: Loop/Managers/LoopDataManager.swift [hunk 25]
```diff
@@ -61,7 +63,12 @@ final class LoopDataManager {
         overrideHistory: TemporaryScheduleOverrideHistory = UserDefaults.appGroup?.overrideHistory ?? .init(),
         lastPumpEventsReconciliation: Date?,
         analyticsServicesManager: AnalyticsServicesManager,
-        localCacheDuration: TimeInterval = .days(1)
+        localCacheDuration: TimeInterval = .days(1),
+        doseStore: DoseStoreProtocol? = nil,
```
Comment (ps2) [inline Loop/Managers/LoopDataManager.swift]: I think the stores can be made non-optional, now that both paths (testing, and normal usage) both provide stores, as well as the backup creation logic below.

File: Loop/Managers/DeviceDataManager.swift [hunk 26]
```diff
@@ -257,6 +336,21 @@ final class DeviceDataManager {
 
         return Manager.init(rawState: rawState) as? CGMManagerUI
     }
+    
+    // Get HealthKit authorization for all of the stores
+    func authorize(_ completion: @escaping () -> Void) {
+        // Authorize all types at once for simplicity
+        carbStore.healthStore.requestAuthorization(toShare: sampleTypes, read: sampleTypes) { (success, error) in
```
Comment (ps2) [inline Loop/Managers/DeviceDataManager.swift]: Can use the local healthStore reference and not have to reach into carbStore

File: Loop/Managers/RemoteDataServicesManager.swift (hunk 20 above)
Comment (ps2) [reply Loop/Managers/RemoteDataServicesManager.swift]: I don't understand why this can't keep using the normal stores?  You don't need to instantiate RemoteDataServicesManager to run LDM tests, right?

File: Loop/View Controllers/CarbAbsorptionViewController.swift (hunk 22 above)
Comment (ps2) [reply Loop/View Controllers/CarbAbsorptionViewController.swift:51]: This is a good cleanup; there are a bunch more that can be cleaned up in this same manner.

Comment (ps2) [reply Loop/View Controllers/CarbAbsorptionViewController.swift:51]: Probably easiest to find by making the store vars on LDM private. <- and that's probably a good change to keep.

**Review by ps2 — CHANGES_REQUESTED** [review]
Looking really good. I think there is just some more cleanup left, which I've left notes on.

File: Loop/Managers/RemoteDataServicesManager.swift (hunk 20 above)
Comment (novalegra) [reply Loop/Managers/RemoteDataServicesManager.swift]: Whoops, I meant to put that comment in LDM. Agreed that this should use `DoseStore`

File: Loop/View Controllers/CarbAbsorptionViewController.swift (hunk 22 above)
Comment (novalegra) [reply Loop/View Controllers/CarbAbsorptionViewController.swift:51]: Wow, there were like 50 references

File: Loop/Managers/LoopDataManager.swift (hunk 25 above)
Comment (novalegra) [reply Loop/Managers/LoopDataManager.swift]: Done! It looks so much cleaner now : )

File: Loop/Managers/RemoteDataServicesManager.swift [hunk 27]
```diff
@@ -126,7 +126,12 @@ extension RemoteDataServicesManager {
             let semaphore = DispatchSemaphore(value: 0)
             let queryAnchor = UserDefaults.appGroup?.getQueryAnchor(for: remoteDataService, withRemoteDataType: .carb) ?? CarbStore.QueryAnchor()
 
-            self.carbStore.executeCarbQuery(fromQueryAnchor: queryAnchor, limit: remoteDataService.carbDataLimit ?? Int.max) { result in
+            guard let carbStore = self.carbStore as? CarbStore else {
```
Comment (ps2) [inline Loop/Managers/RemoteDataServicesManager.swift]: This isn't needed anymore.

**Review by ps2 — APPROVED** [review]
LGTM!  Just one tiny cleanup, and no additional review needed after that. This is great!

File: Loop/Managers/RemoteDataServicesManager.swift [hunk 28]
```diff
@@ -125,7 +125,7 @@ extension RemoteDataServicesManager {
         dispatchQueue(for: remoteDataService, withRemoteDataType: .carb).async {
             let semaphore = DispatchSemaphore(value: 0)
             let queryAnchor = UserDefaults.appGroup?.getQueryAnchor(for: remoteDataService, withRemoteDataType: .carb) ?? CarbStore.QueryAnchor()
-
+
```
Comment (darinkrauss) [inline Loop/Managers/RemoteDataServicesManager.swift:128]: [nit] I try to avoid unnecessary whitespace changes since it just makes the relative delta larger. Don't worry about backing it out now, but consider catching them when doing your final review before committing or PR submit.

File: Loop/Managers/WatchDataManager.swift [hunk 29]
```diff
@@ -168,13 +168,13 @@ final class WatchDataManager: NSObject {
...
         let loopManager = deviceManager.loopManager!
 
-        let glucose = loopManager.glucoseStore.latestGlucose
-        let reservoir = loopManager.doseStore.lastReservoirValue
+        let glucose = deviceManager.glucoseStore.latestGlucose
+        let reservoir =  deviceManager.doseStore.lastReservoirValue
         let basalDeliveryState = deviceManager.pumpManager?.status.basalDeliveryState
 
-        loopManager.getLoopState { (manager, state) in
+        loopManager.getLoopState { [weak deviceManager] (manager, state) in
```
Comment (darinkrauss) [inline Loop/Managers/WatchDataManager.swift]: I think switching to a `weak deviceManager` is going to cause problems in this function. For example, there are a number of `updateGroup.enter()` invocations that will *not* be matched with a corresponding `updateGroup.leave()` if `deviceManager` is `nil` (since the functions with completion handlers that eventually call `updateGroup.leave()` will never be called).

Plus, if `deviceManager` is `nil`, then the end result will be a mostly incomplete (and likely useless) context.

I think you need a guard at the top to check if deviceManager is `nil` (and, perhaps, get a strong reference to do away with all of the `?` below). However, you'll have to consider what you do in this guard. The completion handler needs to be called, but returning an empty `WatchContext` may not be the best choice. Might need to update the completion handler to accept an *optional* `WatchContext` and let downstream worry about a missing context.

@ps2 Thoughts?

File: Loop/View Controllers/StatusTableViewController.swift [hunk 30]
```diff
@@ -329,7 +329,7 @@ final class StatusTableViewController: LoopChartsTableViewController {
         
         // TODO: Don't always assume currentContext.contains(.status)
         reloadGroup.enter()
-        self.deviceManager.loopManager.getLoopState { (manager, state) -> Void in
+        self.deviceManager.loopManager.getLoopState { [weak deviceManager] (manager, state) -> Void in
```
Comment (darinkrauss) [inline Loop/View Controllers/StatusTableViewController.swift]: Same problem here with `weak deviceManager` as in `Loop/Managers/WatchDataManager.swift`. See that file for long comment.

File: Loop/Managers/LoopDataManager.swift [hunk 31]
```diff
@@ -61,7 +63,12 @@ final class LoopDataManager {
         overrideHistory: TemporaryScheduleOverrideHistory = UserDefaults.appGroup?.overrideHistory ?? .init(),
         lastPumpEventsReconciliation: Date?,
         analyticsServicesManager: AnalyticsServicesManager,
-        localCacheDuration: TimeInterval = .days(1)
+        localCacheDuration: TimeInterval = .days(1),
+        doseStore: DoseStoreProtocol,
+        glucoseStore: GlucoseStoreProtocol,
+        carbStore: CarbStoreProtocol,
+        cacheStore: PersistenceController = PersistenceController.controllerInAppGroupDirectory(),
```
Comment (darinkrauss) [inline Loop/Managers/LoopDataManager.swift]: [nit] Not sure if I agree with using a default here. Specifying a default implies (at least to me) that using a different PersistenceController than the one used by the dose, glucose, or carb stores is okay (when it is not). I would remove the default here.

File: Loop/Managers/LoopDataManager.swift [hunk 32]
```diff
@@ -795,7 +736,7 @@ extension LoopDataManager {
         // Fetch glucose effects as far back as we want to make retroactive analysis
         var latestGlucoseDate: Date?
         updateGroup.enter()
-        glucoseStore.getCachedGlucoseSamples(start: Date(timeIntervalSinceNow: -settings.inputDataRecencyInterval)) { (values) in
+        glucoseStore.getCachedGlucoseSamples(start: Date(timeIntervalSinceNow: -settings.inputDataRecencyInterval), end: nil) { (values) in
```
Comment (darinkrauss) [inline Loop/Managers/LoopDataManager.swift]: I *think* you are going to need to do a search for all `Date()` and `timeIntervalSinceNow` and replace with appropriate versions using `now()`.

File: Loop/Managers/LoopDataManager.swift [hunk 33]
```diff
@@ -821,7 +762,7 @@ extension LoopDataManager {
         if insulinEffect == nil {
             self.logger.debug("Recomputing insulin effects")
             updateGroup.enter()
-            doseStore.getGlucoseEffects(start: nextEffectDate) { (result) -> Void in
+            doseStore.getGlucoseEffects(start: nextEffectDate, end: nil, basalDosingEnd: now()) { (result) -> Void in
```
Comment (darinkrauss) [inline Loop/Managers/LoopDataManager.swift:762]: Also, in relation to `now()`, I think you are going to need to look for any default arguments that use `= Date()` throughout.

File: Loop/Managers/LoopDataManager.swift [hunk 34]
```diff
@@ -1896,23 +1842,32 @@ extension LoopDataManager {
         guard FeatureFlags.simulatedCoreDataEnabled else {
             fatalError("\(#function) should be invoked only when simulated core data is enabled")
         }
+        
+        guard let doseStore = doseStore as? DoseStore else {
```
Comment (darinkrauss) [inline Loop/Managers/LoopDataManager.swift]: [nit] I would copy the pattern you use above in `generateSimulatedHistoricalCoreData` to check all of the stores at the top of the function and call `fatalError` once if any fail.

File: Loop/Managers/LoopDataManager.swift [hunk 35]
```diff
@@ -1896,23 +1842,32 @@ extension LoopDataManager {
...
 
-        self.doseStore.purgeHistoricalPumpEvents() { error in
+        doseStore.purgeHistoricalPumpEvents() { error in
             guard error == nil else {
                 completion(error)
                 return
             }
             self.dosingDecisionStore.purgeHistoricalDosingDecisionObjects() { error in
-                guard error == nil else {
+                guard error == nil, let carbStore = self.carbStore as? CarbStore else {
```
Comment (darinkrauss) [inline Loop/Managers/LoopDataManager.swift]: [nit] Relative to the preceding comment, why is carbStore handled specially? That is, no `fatalError` here? Consider previous recommendation of following pattern in `generateSimulatedHistoricalCoreData`.

**Review by darinkrauss — CHANGES_REQUESTED** [review]
Several comments with a couple that probably need to be addressed. If you or @ps2 think it is fine as-is, please let me know and I'll approve. Thanks!

File: Loop/Managers/WatchDataManager.swift (hunk 29 above)
Comment (ps2) [reply Loop/Managers/WatchDataManager.swift]: Good catch, Darin. @novalegra what was the reason to switch to weak deviceManager?

File: Loop/Managers/LoopDataManager.swift (hunk 31 above)
Comment (ps2) [reply Loop/Managers/LoopDataManager.swift]: Might suggest that cacheStore is removed completely, and settingsStore and DosingDecisionStores are constructed by DDM as well, and passed into LDM.

File: Loop/Managers/WatchDataManager.swift (hunk 29 above)
Comment (novalegra) [reply Loop/Managers/WatchDataManager.swift]: I was trying to avoid a retain loop, but that might not be a concern here?

Comment (ps2) [reply Loop/Managers/WatchDataManager.swift]: Yeah, I don't see a retain loop here.

Comment (novalegra) [reply Loop/Managers/WatchDataManager.swift]: I'll revert back to a strong reference

File: Loop/Managers/LoopDataManager.swift (hunk 31 above)
Comment (novalegra) [reply Loop/Managers/LoopDataManager.swift]: Yep, I removed cacheStore from LDM and constructed settingsStore and dosingdecisionStore in DDM

**Review by darinkrauss — APPROVED** [review]
LGTM! :shipit:

**Review by ps2 — APPROVED** [review]
LGTM!
